### PR TITLE
Helm test prometheus

### PR DESCRIPTION
**Pull request by jtudelag**
## Describe the behavior changes introduced in this PR

This PR adds a helm test hook in the form of a k8s job.
This job checks all the Prometheus targets are "UP".

To run the test once Pelorus has been deployed succesfully:
`helm test pelorus`

If this get merged, I would like to reactor testing documentation with this update, as it seems a bit disperse/hidden in the repo.

A couple of things to look into this PR:

- The container image i am using is from docker, it requites curl and jq
- I did bumped chart version manually, as it seems the pre-commit is not working in my laptop.
 

## Linked Issues?

related to #https://github.com/redhat-cop/pelorus/issues/36 

## Testing Instructions

Please include any additional commands or pointers in addition to our [standard PR testing process](/docs/Development.md#testing-pull-requests).

@redhat-cop/mdt

### CODE REVIEW

**Comment by themoosman** [comment]
@jtudelag Can you rebase your branch?

**Review by themoosman — CHANGES_REQUESTED** [review]
Can you rebase the branch.

**Comment by jtudelag** [comment]
@themoosman done.

Still, The container image I am using is from docker hub, it requites curl and jq. Given the recent docker hub registry pull restrictions....Would it make sense to have a pelorus-test image with all the "testing" tools included? helm, bats, conftest, etc ,etc.
What do you think?

**Comment by jtudelag** [comment]
Actually sth similar to this would work: https://github.com/redhat-cop/github-actions/blob/master/confbatstest/Dockerfile

**Comment by themoosman** [comment]
```
[kmoos 🎩︎pelorus-jtudelag]$ helm test pelorus
Error: release: not found
```

**Review by themoosman — CHANGES_REQUESTED** [review]
When running the test command, it never returns.  It just creates a job after job and no return code appears.

**Comment by jtudelag** [comment]
@themoosman So can you paste the logs here? Are your prometheus targets up? Is any of them down? How many prometheus targets you have?

File: charts/pelorus/templates/tests/pelorus-prometheus-test-job.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+spec:
+  parallelism: 1
+  completions: 1
+  template:
+    metadata:
+      name: pelorus-prometheus-test
+    spec:
+      containers:
+      - name: prometheus-targets-test
+        image: docker.io/dwdraju/alpine-curl-jq # TODO: decide  what image to use...
```
Comment (themoosman) [inline charts/pelorus/templates/tests/pelorus-prometheus-test-job.yml:17]: This might not work in some client environments.

**Comment by themoosman** [comment]
> @themoosman So can you paste the logs here? Are your prometheus targets up? Is any of them down? How many prometheus targets you have?

What logs?  The cli command never returns, it just hangs.

**Comment by jtudelag** [comment]
> > @themoosman So can you paste the logs here? Are your prometheus targets up? Is any of them down? How many prometheus targets you have?
> 
> What logs? The cli command never returns, it just hangs.

You mention this in a previous comment:

`When running the test command, it never returns. It just creates a job after job and no return code appears.`

So I was asking for the logs of that jobs.

File: charts/pelorus/templates/tests/pelorus-prometheus-test-job.yml (hunk 1 above)
Comment (jtudelag) [reply charts/pelorus/templates/tests/pelorus-prometheus-test-job.yml:17]: Please see my previous comment https://github.com/redhat-cop/pelorus/pull/227#issuecomment-721002502

Comment (themoosman) [reply charts/pelorus/templates/tests/pelorus-prometheus-test-job.yml:17]: Could you use a pattern similar to this?  https://github.com/etsauer/pelorus/blob/grafana-snapshots/charts/pelorus/templates/dashboard-snapshot-cronjob.yaml#L58

Comment (jtudelag) [reply charts/pelorus/templates/tests/pelorus-prometheus-test-job.yml:17]: Yes, I will do. On the other hand, that approach won't work on disconnected environments.

Comment (themoosman) [reply charts/pelorus/templates/tests/pelorus-prometheus-test-job.yml]: I've created issue #237 to address the image issue.  At least now we're using the same pattern across all the jobs.

**Comment by themoosman** [comment]
There is still something odd going on.  The jobs keeps failing even when the endpoints are healthy.  

So I added a debug line as in:

```
...
# Test all Prom targets are "UP"
          echo -e "\n Checking if all targets are UP...\n"
          for t in "${_PROM_TARGETS_HEALTH}"
          do
              echo "$t"
              if [[ "$t" != "up" ]]
              then
                  echo -e "Test failed, one target is not UP.\n"
                  exit -1
              fi

          done
...

```

and got the following:
```
Checking Prometheus Targets Health.

Executing tests inside a Kubernetes Pod.

Final URL: http://prometheus-pelorus:9090/api/v1/targets

*   Trying 172.30.241.213...
* TCP_NODELAY set
* Connected to prometheus-pelorus (172.30.241.213) port 9090 (#0)
> GET /api/v1/targets HTTP/1.1
> Host: prometheus-pelorus:9090
> User-Agent: curl/7.61.1
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Type: application/json
< Date: Thu, 12 Nov 2020 15:17:36 GMT
< Transfer-Encoding: chunked
<
{ [10720 bytes data]
* Connection #0 to host prometheus-pelorus left intact

 Checking if all targets are UP...

up
up
Test failed, one target is not UP.
```

You can see the two `up` statements, but the job still fails.

**Comment by jtudelag** [comment]
Hey @themoosman thanks for the review. The test was broken because of the quotes in "${_PROM_TARGETS_HEALTH}".

I have refactored it a bit to make it a bit more verbose, so when a target is not up, we know which one.

Hopefully now works...

**Comment by themoosman** [comment]
The test is now working just fine.  I did notice one last thing.  When doing a `helm uninstall` the `pelorus-prometheus-test` job is not removed.

**Comment by jtudelag** [comment]
I would say this is helm expected behaviour, as [helm uninstall](https://helm.sh/docs/helm/helm_uninstall/) applies to the helm release, and the tests are not part of the release.

I don't see every user installing Pelorus running the tests, just maybe if sth goes wrong, and I would say the main use case would be development purposes and CI/CD (integration tests?)?

Other option is use a hook deletion policy, as test is considered anothe rhook in helm: https://helm.sh/docs/topics/charts_hooks/#hook-deletion-policies

**Comment by etsauer** [comment]
@themoosman @jtudelag couple of things. 

1. I agree that the cleanup is not super important, as this would mostly be used for CI purposes. However, from the annotations, it looks like we should expect it to be cleaned up like a hook. If we're going to have those annotations, they should work.
2. Tests aren't valuable unless they get run. And right now this PR does not include either automatic execution of this test, or instructions on running it. Without one of those, this code becomes tech debt. Let's make sure any tests we have in this repo also have a process in place for using them.

**Comment by jtudelag** [comment]
@etsauer 

>     1. I agree that the cleanup is not super important, as this would mostly be used for CI purposes. However, from the annotations, it looks like we should expect it to be cleaned up like a hook. If we're going to have those annotations, they should work.

So, the current annotation `before-hook-creation` work as expected. The question is if we want to add any additional [deletion policy](https://helm.sh/docs/topics/charts_hooks/#hook-deletion-policies) annotation or not. I would argue that we don't want any deletion policy so we can always see if the test went ok or failed. Same behaviour happens to OCP Pod deployments for DCs. the Pod is always kept ;). 

 
>     2. Tests aren't valuable unless they get run. And right now this PR does not include either automatic execution of this test, or instructions on running it. Without one of those, this code becomes tech debt. Let's make sure any tests we have in this repo also have a process in place for using them.

I would rather document this first, and then add any sort of automation in a different PR later.

**Comment by etsauer** [comment]
@jtudelag that's fine, and I agree with just documenting, and not integrating into tests yet (the reality is we can't test this automatically, because we don't yet have openshift environments to CI against). My point was that this PR should include updates to the developer docs to instruct developers/PR reviewers to actually USE this test.

**Comment by etsauer** [comment]
@jtudelag looking at your README updates, it looks like you've made a lot of changes that have nothing to do with this new test. Could you clean that up please?

**Comment by jtudelag** [comment]
@etsauer it is all MD formatting improvements/fixes, but ok.

**Comment by etsauer** [comment]
@jtudelag this looks good, just need to bump the Chart version again.

**Review by etsauer — APPROVED** [review]

**Comment by jtudelag** [comment]
@etsauer @themoosman  PTAL

**Comment by mpryc** [comment]
I am closing this PR as a result of current cleanup of open issues/PRs.

Helm testing is still valid scenario, however the way we would want to implement this have changed, because we are creating Prometheus and Grafana deployments via OLM dependency in the Pelorus operator.

**Pull request closed without merging**